[PR] Adding proxy contract for open source algoworld swapper

aorumbayev opened the pull request:
### Description

- Adds a proxy account that serves as a reference point for obtaining swap configurations for users of swapper.algoworld.io. The proxy contract accepts 2 pay transactions, one covering the fee for the consecutive transaction that is issued on behalf of proxy. The second transaction is always a 0 amount pay transaction who's note field contains the link to ipfs url that contains the swapper configuration file ([here](https://github.com/AlgoWorldNFT/algoworld-swapper/blob/main/api/storage/saveConfigurations.ts) is a reference on the function that stores configs in ipfs in swapper). @cusma let me know if you have any questions. As usual if you see that i merged the PR already just leave the comments on it regardless (if you have any) - and i will address in timely manner.
- Adds workflow step that pushed the package to `pypi`

### Checklist

> _Please, make sure to comply with the checklist below before expecting review_

- [x] Code compiles correctly
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

aorumbayev commented on algoworld_contracts/swapper/swap_proxy.py:
```diff
@@ -0,0 +1,119 @@
+"""
+MIT License
```
@cusma  here is the new proxy contract that was added.

cusma commented on algoworld_contracts/swapper/swap_proxy.py:
```diff
@@ -0,0 +1,119 @@
+
+
+def proxy_store(cfg: SwapProxy):
+    Balance(Txn.sender())
```
What's the scope of `Balance` here?

cusma commented on algoworld_contracts/swapper/swap_proxy.py:
```diff
@@ -0,0 +1,119 @@
+    store_fee = And(
+        Gtxn[STORE_FEE].type_enum() == TxnType.Payment,
+        Or(
+            Gtxn[STORE_FEE].amount() == Int(110_000),
+            Gtxn[STORE_FEE].amount() == Int(10_000),
```
Doubt: why should a user pay `110_000` if paying just `10_000` is approved by the contract?

aorumbayev replied:
@cusma pooled fees sounds cool actually wasnt aware of it. I did wanted to have it atomic though to minimize the amount of different transaction groups/transactions that id force user to sign. In this case this 110_000 really only happens once for any wallet doing first ever swapper, afterwards UI just checks for the state. But i agree this can be decoupled and i could actually tie it with the wallet during first logging and user would have a prompt popup asking to sign proxy config transaction to be able to use swapper (that's also where i can put terms & conditions checkbox). 

I am a bit tied with the release plan though. As usual I will move your comment to github issue and will get back to it once time allows. Thanks for the suggestion as usual!

aorumbayev replied:
Resolving the issue (as the request for change is out of scope for initial release)

cusma commented on algoworld_contracts/swapper/swap_proxy.py:
```diff
@@ -0,0 +1,119 @@
+            Gtxn[STORE_FEE].amount() == Int(10_000),
+        ),
+        Gtxn[STORE_FEE].sender() == Addr(cfg.swap_creator),
+        Gtxn[STORE_FEE].receiver() == Txn.sender(),
```
I'm not sure about what is happening here: can you give a description of this step? Who is paying who, for what? Who is executing the `STORE_FEE` transaction? Is that a user or the Smart Contract?

aorumbayev replied:
Perhaps `Txn.sender(),` is not used in the right context here but this transaction is the one that creator of the proxy pays to cover the fee for the second transaction that issued on behalf of the escrow (with logisig) that sends 0 algos to itself and puts ipfs note in there. 

With `Gtxn[STORE_FEE].receiver() == Txn.sender()` the intent was to check if the receiver is the escrow account, since tests werent failing i assume  Txn.sender() in this context does indeed return me the address of the escrow?

aorumbayev replied:
Also, pushed a commit that removes them already

cusma commented on algoworld_contracts/swapper/swap_proxy.py:
```diff
@@ -0,0 +1,119 @@
+        Gtxn[STORE_PROXY_NOTE].amount() == Int(0),
+        Gtxn[STORE_PROXY_NOTE].sender() == Txn.sender(),
+        Gtxn[STORE_PROXY_NOTE].receiver() == Txn.sender(),
+        Gtxn[STORE_PROXY_NOTE].rekey_to() == Global.zero_address(),
+        Gtxn[STORE_PROXY_NOTE].close_remainder_to() == Global.zero_address(),
```
Seems that `STORE_PROXY_NOTE` is an user zero Payment self-transaction. Since the transaction in not signed by the Smart Contract the check on `rekey_to` and `close_remainder_to` are not necessary. Am I understanding it correctly?

aorumbayev replied:
Yes, that's correct. Wasn't aware that rekey and close_remainder are not necessary in this case. Could you expand on why it isn't needed?

cusma replied:
Both `close_remainder_to` and `rekey_to` take effect on who _signs_ the transaction, turning that into "dangerous" field when is actually the Smart Contract to sign and authorize the transaction. If, like in this case, the signer of the transaction is the user it would not be "dangerous" if the user wants to rekey or close its own account with this transaction, those are effects involving user's account (not the Smart Contract).

aorumbayev commented on tests/swapper/test_swap_proxy.py:
```diff
@@ -0,0 +1,64 @@
+    )
+
+
+def test_swap_proxy_activate_or_save(swap_creator, swap_proxy):
```
@cusma  you can also see the tests covering the proxy contract here, its really only used for dumping the ipfs urls

aorumbayev commented:
@cusma hey - thanks for review, i addressed most of the comments but for your suggestion to decouple funding transaction of 110_000 algos, ill address it in separate pr https://github.com/AlgoWorldNFT/algoworld-contracts/issues/48

cusma reviewed (approved):
LGTM
